src/sentry/search/django/backend.py

```diff
@@ -8,286 +8,445 @@
+            )
 
-            if query:
-                event_queryset = event_queryset.filter(
-                    message__icontains=query)
+            get_sort_expression, sort_value_to_cursor_value = environment_sort_strategies[sort_by]
 
-            # limit to the first 1000 results
-            group_ids = event_queryset.distinct().values_list(
-                'group_id', flat=True)[:1000]
+            group_tag_value_queryset = tagstore.get_group_tag_value_qs(
+                project.id,
+                set(group_queryset.values_list('id', flat=True)),  # TODO: Limit?,
+                environment.id,
+                'environment',
+                environment.name,
+            )
 
-            # if Event is not on the primary database remove Django's
-            # implicit subquery by coercing to a list
-            base = router.db_for_read(Group)
-            using = router.db_for_read(Event)
-            # MySQL also cannot do a LIMIT inside of a subquery
-            if base != using or engine.startswith('mysql'):
-                group_ids = list(group_ids)
+            candidates = dict(
+                QuerySetBuilder({
+                    'age_from': ScalarCondition('first_seen', 'gt'),
+                    'age_to': ScalarCondition('first_seen', 'lt'),
+                    'last_seen_from': ScalarCondition('last_seen', 'gt'),
+                    'last_seen_to': ScalarCondition('last_seen', 'lt'),
+                    'times_seen': CallbackCondition(
+                        lambda queryset, times_seen: queryset.filter(times_seen=times_seen),
+                    ),
+                    'times_seen_lower': ScalarCondition('times_seen', 'gt'),
+                    'times_seen_upper': ScalarCondition('times_seen', 'lt'),
+                }).build(
+                    group_tag_value_queryset,
+                    parameters,
+                ).extra(
+                    select={
+                        'sort_value': get_sort_expression(group_tag_value_queryset.model),
+                    },
+                ).values_list('group_id', 'sort_value')
+            )
 
-            queryset = queryset.filter(
-                id__in=group_ids,
+            if tags:
+                matches = tagstore.get_group_ids_for_search_filter(
+                    project.id,
+                    environment.id,
+                    tags,
+                    candidates.keys(),
+                    limit=len(candidates),
+                )
+                for key in set(candidates) - set(matches or []):
+                    del candidates[key]
+
+            result = SequencePaginator(
+                map(
+                    lambda (id, score): (sort_value_to_cursor_value(score), id),
+                    candidates.items(),
+                ),
+                reverse=True,
+                **paginator_options
+            ).get_result(limit, cursor, count_hits=count_hits)
+
+            result.results = filter(
```

Comment:
same here, filter returns an iterator in python 3, so if it needs to be a list we hsould use a list comprehension

Reply:
Will change.

Reply:
Updated here: 6c4c2ef